// linn-cloud-libs/config/index.js
"use strict";

var winston = require('winston');

var currentEnvironment = process.env.NODE_ENV;

if (!currentEnvironment) {
    throw(new Error("No environment set up"));
}

switch(currentEnvironment) {
    case "test":
        module.exports = {dynamoDb: { tables: {} }};
        break;
    case "debug":
        module.exports = require('../../../config/debug.js');
        break;
    case "int":
        module.exports = require('../../../config/int.js');
        break;
    case "release":
        module.exports = require('../../../config/release.js');
        break;
    default:
        throw(new Error("No configuration for " + currentEnvironment));
}

module.exports.requestLoggerFormat = ':remote-addr :method :url :status :response-time';

var logger = new winston.Logger({
    transports: [
        new winston.transports.Console({
            timestamp: function() {
                return new Date().toISOString();
            },
            level: 'debug',
            handleExceptions: false,
            json: false,
            colorize: false
        })
    ]
});
logger.stream = {
    write: function(message, encoding) {
        logger.info(message);
    }
};

module.exports.logger = logger;
