// linn-cloud-libs/dynamodb/repositories/cloudProductDescriptorRepository.js
"use strict";

var aws = require('aws-sdk');
var doc = require('dynamodb-doc');
var config = require('../../config');

aws.config.region = config.awsRegion;
var docClient = new doc.DynamoDB();

exports.findById = function loadCloudProductDescriptorByIdFromDynamoDb(productDescriptorId, callback) {
    var params = {
        TableName: config.dynamoDb.tables.productDescriptors,
        Key : {
            id: productDescriptorId
        }
    };
    docClient.getItem(params,  function(err, results) {
        if (err) {
            callback(err);
        } else {
            callback(null, results.Item);
        }
    });
};

exports.filterBy = function loadCloudProductDescriptorsFromDynamoDb(vendor, productType, callback) {
    var params = {
        TableName: config.dynamoDb.tables.productDescriptors,
        IndexName: config.dynamoDb.indices.productDescriptors,
        KeyConditions : [
            docClient.Condition("vendor", "EQ", vendor),
            docClient.Condition("productType", "EQ", productType)
        ]
    };
    docClient.query(params, function(err, results) {
        if (err) {
            callback(err);
        } else {
            callback(null, results.Items);
        }
    });
};

exports.addOrReplace = function saveCloudProductDescriptorToDynamoDb(cloudProductDescriptor, callback) {
    var params = {
        TableName: config.dynamoDb.tables.productDescriptors,
        Item: cloudProductDescriptor
    };
    docClient.putItem(params, callback);
};

exports.removeById = function removeCloudProductDescriptorFromDynamoDb(productDescriptorId, callback) {
    var params = {
        TableName: config.dynamoDb.tables.productDescriptors,
        Key : {
            id: productDescriptorId
        }
    };
    docClient.deleteItem(params, function (err) {
        if (err) {
            callback(err);
        } else {
            callback();
        }
    })
};
