// linn-cloud-libs/dynamodb/repositories/cloudRepository.js
"use strict";

var aws = require('aws-sdk');
var doc = require('dynamodb-doc');
var config = require('../../config');

aws.config.region = config.awsRegion;
var docClient = new doc.DynamoDB();

module.exports.CloudRepository = function CloudRepository(tableName, hashKey) {
    function findByIdFromDynamoDb(id, callback) {
        var params = {
            TableName: tableName,
            Key: {}
        };
        params.Key[hashKey] = id;
        this.docClient.getItem(params, function (err, result) {
            if (err) {
                callback(err);
            } else {
                callback(null, result.Item);
            }
        });
    }
    function addOrReplaceByIdToDynamoDb(item, callback) {
        var params = {
            TableName: tableName,
            Item: item
        };
        this.docClient.putItem(params, callback);
    }
    function removeByIdFromDynamoDb(id, callback) {
        var params = {
            TableName: tableName,
            Key: {}
        };
        params.Key[hashKey] = id;
        this.docClient.deleteItem(params, function (err) {
            if (err) {
                callback(err);
            } else {
                callback();
            }
        })
    }
    return {
        docClient: docClient,
        findById: findByIdFromDynamoDb,
        addOrReplace: addOrReplaceByIdToDynamoDb,
        removeById: removeByIdFromDynamoDb
    };
};
